### PR TITLE
RPC for ACL managment

### LINKED ISSUE #4363: RPC for ACL management
## Rationale

Currently, there are only operations for adding block rule to ACL: `net.peer.block`, 
`net.peer.block_ip`. But there is no way to:
 - list current rules
 - remove node from blacklist.

### Functional specification

New operations:

| uri | name | description |
| -- | -- | -- |
| net.peer.allow | allow_node | Whitelists node |
| net.peer.allow_ip | allow_ip | Removes ip from blacklist |
| net.peer.acl | acl_status | Returns ACL configuration for nodes |
| net.peer.acl_ip | acl_ip_status | Returns ACL for ip addresses | 
| net.peer.acl.new | acl_setup | Resets nodes ACL to new configuration |

Existing operations

| uri | name | description |
| -- | -- | -- |
| net.peer.block | disallow_node | Blacklists node |
| net.peer.block_ip | disallow_ip | Adds ip to blacklist |

#### `allow_node`

**URI**: `net.peer.allow`

| In/Out | name | type | is optional | 
| -- | -- | -- | -- |
| IN | node_id | str | No |
| IN | persist | bool | Yes, default `true` |
| OUT | - | None | |

#### `allow_ip`

**URI**: `net.peer.allow_ip`

| In/Out | name | type | is optional | 
| -- | -- | -- | -- |
| IN | ip | str | No |
| IN | persist | bool | Yes, default `true` |
| OUT | - | None | |

#### acl_status

**URI**: `net.peer.acl`

| In/Out | name | type | is optional | 
| -- | -- | -- | -- |
| OUT | AclStatus |  | |

#### acl_ip_status

**URI**: `net.peer.acl_ip`

| In/Out | name | type | is optional | 
| -- | -- | -- | -- |
| OUT | AclStatus |  | |

#### acl_setup

**URI**: `net.peer.acl.new`

| In/Out | name | type | description | 
| -- | -- | -- | -- |
| IN | default_rule | str | required, valid values: `allow`, `deny` |
| IN | exceptions | `List[str]` | required, list of exceptions to default rule |

#### AclStatus

| Field | type | description | 
| default_rule | str | enum: `allow`, `deny` |
| rules | List[Tuple[str, str, Optional[int]]] | list of triplets in form: `(item, rule, duration)`. |

**Examples:**

```json
{"default_rule": "allow",
    "rules": [
      [
        "8c6c3bcb42a1a7a555118a14c737e06aac16e17045523cef7c61facd17a8eeff971f377518773848312f312aaf1a2d8b3709376106cb3f6e4acbcf47a0cb9361",
        "deny",
        null
      ]
   ]
}
```

```json
{
"default_rule": "allow",
"rules":[
    [
      "10.30.10.103",
      "deny",
      3121874772.651
    ],
    [
      "10.30.10.184",
      "deny",
      3121874794.917
    ]
  ]
}
```

## Dependencies

* Featue is related to: golemfactory/golem-client#21

## QA
_Description of the QA process for the feature_

#### Base test process
_specify one of: ( if unsure, consult @ZmijaWA or @ederenn )_
** QA limited to the single feature (in case of very trivial changes)
** Smoke Test (the limited set of tests designed to ensure that most critical functionality doesn't experience regressions)
** Full Test (full set of tests for the whole application - in case of more extensive application updates, refactoring etc)

### New feature tests
_Add test scenarios for the new feature_

#### Positive Scenario 1
* Step 1
* Step 2
* ...
* Expected results

#### Positive Scenario 2
* ...

#### Negative Scenario 1
* Step 1
* Step 2
* ...
* Some expected problem
* Expected problem handling

### Possible regressions
* is there any functionality that is likely to experience regressions?
* are there any existing application components that should be tested more thoroughly?

### Development
* [x] Feature implemented
* [x] Feature unit/integration tests implemented

### Dev QA
* [x] Basic tests by the developer pass
* [x] Unit tests pass

### QA team
* [ ] Base scenario passes
* [ ] Additional test 1 passes... 
* [ ] ... 

## Issues encountered during QA
_When adding an issue here, please update testing scenarios and QA progress above_
* [ ] link to issue... 
* [ ] ...

### PR DESCRIPTION
**Pull request by prekucki**
resolves #4363

### CODE REVIEW

File: golem/task/acl.py [hunk 1]
```diff
@@ -31,6 +57,14 @@ def disallow(self, node_id: str, timeout_seconds: int, persist: bool) \
             -> None:
         pass
 
+    @abc.abstractmethod
+    def allow(self, node_id: str, persist: bool) -> None:
+        pass
```
Comment (tworec) [inline golem/task/acl.py:62]: raise NotImplemented
```suggestion
        raise NotImplementedError
```
see https://github.com/golemfactory/golem/pull/4455
and https://docs.python.org/3/library/exceptions.html#NotImplementedError

File: golem/task/acl.py [hunk 2]
```diff
@@ -31,6 +57,14 @@ def disallow(self, node_id: str, timeout_seconds: int, persist: bool) \
...
+
+    @abc.abstractmethod
+    def status(self) -> AclStatus:
+        pass
```
Comment (tworec) [inline golem/task/acl.py:66]: ditto

File: golem/task/taskserver.py [hunk 3]
```diff
@@ -832,6 +832,29 @@ def disallow_node(self, node_id: str, timeout_seconds: int, persist: bool) \
...
+    @rpc_utils.expose('net.peer.acl')
+    def acl_status(self):
+        return self.acl.status().to_message()
+
+    @rpc_utils.expose('net.peer.acl_ip')
+    def acl_ip_status(self):
+        return self.acl_ip.status().to_message()
+
+    @rpc_utils.expose('net.peer.acl.new')
+    def acl_setup(self, default_rule: str, exceptions: List[str]) -> None:
```
Comment (tworec) [inline golem/task/taskserver.py:852]: what about `acl_ip_setup`. Seems to be missing...

**Review by tworec — DISMISSED** [review]

**Comment by shadeofblue** [comment]
@prekucki @tworec general question - why are we using a textfile instead of the db for persistence here?

**Comment by prekucki** [comment]
> @prekucki @tworec general question - why are we using a textfile instead of the db for persistence here?

It is question to your team. i just added rpc endpoints to read and modify this data.

File: golem/task/acl.py (hunk 1 above)
Comment (shadeofblue) [reply golem/task/acl.py:62]: I'd agree with @tworec here

**Comment by shadeofblue** [comment]
> > @prekucki @tworec general question - why are we using a textfile instead of the db for persistence here?
> 
> It is question to your team. i just added rpc endpoints to read and modify this data.

well, I don't this it's something to do with _our_ team specifically but sure, I understand that it's a preexisting condition..

**Comment by shadeofblue** [comment]
@prekucki anyway, will you be implementing the changes suggested by @tworec ?

**Review by shadeofblue — APPROVED** [review]

File: golem/task/acl.py (hunk 1 above)
Comment (tworec) [reply golem/task/acl.py:62]: I'll add it in #4588

**Pull request merged**